src/main/java/org/wildfly/security/credential/storage/CredentialStorageSpi.java

```diff
@@ -0,0 +1,129 @@
+ */
+public abstract class CredentialStorageSpi {
+
+    /**
+     * Field indicating successful initialization ({@link #initialize(Map)}. Each subclass should set this field.
+     */
+    protected boolean initialized = false;
+
+    /**
+     * Construct a new instance of this SPI.
+     */
+    protected CredentialStorageSpi() {
+    }
+
+    /**
+     * Initialize credential storage service with given attributes. This procedure should set {@link #initialized} after
+     * successful initialization.
+     *
+     * @param attributes attributes to used to pass information to credential storage service
+     * @throws StorageException if initialization fails due to any reason
+     */
+    public abstract void initialize(Map<String, String> attributes) throws StorageException;
+
+    /**
+     * Checks whether underlying credential storage service is initialized.
+     *
+     * @return {@code true} in case of initialization passed successfully, {@code false} otherwise.
+     */
+
+    public boolean isInitialized() {
+        return initialized;
+    }
+
+    /**
+     * Check if credential storage service supports modification of its storage
+     * @return {@code true} in case of modification of the storage is supported, {@code false} otherwise
+     */
+    public abstract boolean isModifiable();
+
+    /**
+     * Check whether credential storage service has an entry associated with the given credential alias of specified credential type.
+     * @param credentialAlias key to check existence
+     * @param credentialType to check existence in the credential storage
+     * @param <C> the class of type to which should be credential casted
+     * @return {@code true} in case key exist in storage otherwise {@code false}
+     * @throws StorageException when there is a problem with credential storage
+     * @throws UnsupportedCredentialTypeException when the credentialType is not supported
+     */
+    public abstract <C extends Credential> boolean exists(String credentialAlias, Class<C> credentialType)
+            throws StorageException, UnsupportedCredentialTypeException;
+
+    /**
+     * Store credential to the storage service under the given alias. If given alias already contains specific credential type type the credential
+     * replaces older one. <em>Note:</em> {@link CredentialStorageSpi} supports storing of multiple entries (credential types) per alias.
+     * Each must be of different credential type.
+     * @param credentialAlias to store the credential to the storage
+     * @param credentialType - credential type to be stored under the credentialAlias to the storage
+     * @param credential instance of {@link Credential} to store
+     * @throws StorageException when the credential cannot be stored.
+     * @throws UnsupportedCredentialTypeException when the credentialType is not supported
+     */
+    public abstract  <C extends Credential> void store(String credentialAlias, Class<C> credentialType, C credential)
```

Comment:
I understand specifying the credential type under the exists/retrieve methods because that can be used to cast-or-null the return value, but this implies that a single alias might have more than one credential, and furthermore that the credential selection is type-based, which can be problematic (is the type an exact key? is it a matching parameter? how do we deal with ambiguity?)

I think if we want to support multiple credentials per alias, we should have a primary and secondary alias string.  Of course that's effectively the same as just having one alias string though.